apps/linear_algebra/src/halide_blas.cpp

```diff
@@ -0,0 +1,331 @@
+#include <string.h>
+#include "halide_blas.h"
+
+#define assert(cond) halide_assert(0,cond)
+
+namespace {
+
+void init_scalar_buffer(const float *x, buffer_t *buff) {
+  memset((void*)buff, 0, sizeof(buffer_t));
+  buff->host = (uint8_t*)const_cast<float*>(x);
+  buff->extent[0] = 1;
+  buff->stride[0] = 1;
+  buff->elem_size = sizeof(float);
+}
+
+void init_scalar_buffer(const double *x, buffer_t *buff) {
+  memset((void*)buff, 0, sizeof(buffer_t));
+  buff->host = (uint8_t*)const_cast<double*>(x);
+  buff->extent[0] = 1;
+  buff->stride[0] = 1;
+  buff->elem_size = sizeof(double);
+}
+
+void init_vector_buffer(const int N, const float *x, const int incx, buffer_t *buff) {
+  memset((void*)buff, 0, sizeof(buffer_t));
+  buff->host = (uint8_t*)const_cast<float*>(x);
+  buff->extent[0] = N;
+  buff->stride[0] = incx;
+  buff->elem_size = sizeof(float);
+}
+
+void init_vector_buffer(const int N, const double *x, const int incx, buffer_t *buff) {
+  memset((void*)buff, 0, sizeof(buffer_t));
+  buff->host = (uint8_t*)const_cast<double*>(x);
+  buff->extent[0] = N;
+  buff->stride[0] = incx;
+  buff->elem_size = sizeof(double);
+}
+
+void init_matrix_buffer(const int M, const int N, const float *A, const int lda, buffer_t *buff) {
+  memset((void*)buff, 0, sizeof(buffer_t));
+  buff->host = (uint8_t*)const_cast<float*>(A);
+  buff->extent[0] = M;
+  buff->extent[1] = N;
+  buff->stride[0] = 1;
+  buff->stride[1] = lda;
+  buff->elem_size = sizeof(float);
+}
+
+void init_matrix_buffer(const int M, const int N, const double *A, const int lda, buffer_t *buff) {
+  memset((void*)buff, 0, sizeof(buffer_t));
+  buff->host = (uint8_t*)const_cast<double*>(A);
+  buff->extent[0] = M;
+  buff->extent[1] = N;
+  buff->stride[0] = 1;
+  buff->stride[1] = lda;
+  buff->elem_size = sizeof(double);
+}
+
+}
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+//////////
+// copy //
+//////////
+
+void hblas_scopy(const int N, const float *x, const int incx,
+                 float *y, const int incy) {
+  buffer_t buff_x, buff_y;
+  init_vector_buffer(N, x, incx, &buff_x);
+  init_vector_buffer(N, y, incy, &buff_y);
+  assert(0 == halide_scopy(&buff_x, &buff_y));
```

Comment:
The constant-on-left comparison style isn't really used anywhere else in Halide, so it looks pretty jarring here.